src/monitoring_service/handlers.py

```diff
@@ -406,67 +408,84 @@ def action_monitoring_triggered_event_handler(event: Event, context: Context) ->
-
     if monitor_request.nonce <= last_onchain_nonce:
         log.info(
             "Another MS submitted the last known channel state", monitor_request=monitor_request
         )
+        return
 
+    latest_block = context.w3.eth.blockNumber
+    last_confirmed_block = latest_block - context.required_confirmations
+    user_address = monitor_request.non_closing_signer
+    user_deposit = get_pessimistic_udc_balance(
+        udc=context.user_deposit_contract,
+        address=user_address,
+        from_block=last_confirmed_block,
+        to_block=latest_block,
+    )
+    context.user_deposit_contract.functions.effectiveBalance(user_address).call()
     if monitor_request.reward_amount < context.min_reward:
         log.info(
             "Monitor request not executed due to insufficient reward amount",
             monitor_request=monitor_request,
             min_reward=context.min_reward,
         )
+        return
+
+    if user_deposit < monitor_request.reward_amount * DEFAULT_PAYMENT_RISK_FAKTOR:
+        log.debug(
+            "User deposit is insufficient -> try monitoring again later",
+            monitor_request=monitor_request,
+            min_reward=context.min_reward,
+        )
+        context.db.upsert_scheduled_event(
+            ScheduledEvent(trigger_block_number=BlockNumber(last_confirmed_block + 1), event=event)
+        )
+        return
 
-    call_monitor = (
+    assert (
         channel.closing_tx_hash is None
-        and monitor_request.nonce > last_onchain_nonce
-        and user_deposit >= monitor_request.reward_amount * DEFAULT_PAYMENT_RISK_FAKTOR
-        and monitor_request.reward_amount >= context.min_reward
-    )
-    if call_monitor:
-        try:
-            # Attackers might be able to construct MRs that make this fail.
-            # Since we execute a gas estimation before doing the `transact`,
-            # the gas estimation will fail before any gas is used.
-            # If we stop doing a gas estimation, a `call` has to be done before
-            # the `transact` to prevent attackers from wasting the MS's gas.
-            tx_hash = TransactionHash(
-                bytes(
-                    context.monitoring_service_contract.functions.monitor(
-                        monitor_request.signer,
-                        monitor_request.non_closing_signer,
-                        monitor_request.balance_hash,
-                        monitor_request.nonce,
-                        monitor_request.additional_hash,
-                        monitor_request.closing_signature,
-                        monitor_request.non_closing_signature,
-                        monitor_request.reward_amount,
-                        monitor_request.token_network_address,
-                        monitor_request.reward_proof_signature,
-                    ).transact({"from": context.ms_state.address})
-                )
+    ), "This MS already monitored this channel. Should be impossible."
```

Comment:
If we receive a new MR after calling `monitor`, should we call `monitor` again with the updated BP? That would increase our gas cost. But otherwise a different MS might do that and we ill get no reward, although we already have spent gas.

Reply:
I guess in general this scenario is not likely.

And what you describe sounds like a loose/loose situation, not sure how we get out of that.

Reply:
See https://github.com/raiden-network/raiden-services/issues/504.